File: backend/services/stigmer-server/pkg/domain/workflow/validation/validator.go
package validation

import (
	"context"
	"fmt"

	"github.com/rs/zerolog/log"
	workflowv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/workflow/v1"
	serverlessv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/workflow/v1/serverless"
	"github.com/stigmer/stigmer/backend/services/stigmer-server/pkg/domain/workflow/converter"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// WorkflowValidator validates a WorkflowSpec and returns a ServerlessWorkflowValidation.
// This is the interface that the pipeline step depends on.
type WorkflowValidator interface {
	Validate(ctx context.Context, spec *workflowv1.WorkflowSpec) (*serverlessv1.ServerlessWorkflowValidation, error)
}

// InProcessValidator validates WorkflowSpec proto entirely in-process without
// Temporal. It converts the spec to CNCF Serverless Workflow DSL YAML, runs
// cross-reference validation, and produces budget warnings.
//
// This replaces the Temporal-based ServerlessWorkflowValidator that dispatched
// validation to the now-deleted Go workflow-runner.
type InProcessValidator struct {
	converter *converter.Converter
}

// NewInProcessValidator creates a new in-process workflow validator.
func NewInProcessValidator() *InProcessValidator {
	return &InProcessValidator{
		converter: converter.NewConverter(),
	}
}

// Validate converts WorkflowSpec proto to CNCF YAML and validates it.
//
// Returns ServerlessWorkflowValidation with:
//   - VALID: Conversion succeeded, no structural errors
//   - INVALID: User error (bad structure, missing fields, invalid references)
//   - FAILED: System error (converter crash)
//
// The yaml field always contains the generated CNCF YAML when conversion
// succeeds (even for INVALID state), as it aids debugging.
func (v *InProcessValidator) Validate(ctx context.Context, spec *workflowv1.WorkflowSpec) (*serverlessv1.ServerlessWorkflowValidation, error) {
	if spec == nil {
		return &serverlessv1.ServerlessWorkflowValidation{
			State:       serverlessv1.ValidationState_FAILED,
			Errors:      []string{"WorkflowSpec cannot be nil"},
			ValidatedAt: timestamppb.Now(),
		}, nil
	}

	log.Debug().Msg("Starting in-process workflow validation")

	// Step 0: Task kind validation (fail fast for unknown kinds)
	if kindErrors := ValidateTaskKinds(spec); len(kindErrors) > 0 {
		return &serverlessv1.ServerlessWorkflowValidation{
			State:       serverlessv1.ValidationState_INVALID,
			Errors:      kindErrors,
			ValidatedAt: timestamppb.Now(),
		}, nil
	}

	// Step 1: Convert proto to CNCF YAML
	yaml, err := v.converter.ProtoToYAML(spec)
	if err != nil {
		log.Error().Err(err).Msg("YAML generation failed")
		return &serverlessv1.ServerlessWorkflowValidation{
			State:       serverlessv1.ValidationState_INVALID,
			Errors:      []string{fmt.Sprintf("Failed to generate YAML: %v", err)},
			ValidatedAt: timestamppb.Now(),
		}, nil
	}

	log.Debug().Int("yaml_length", len(yaml)).Msg("YAML generation succeeded")

	var errors []string
	var warnings []string

	// Step 2: Cross-reference validation (unique names, flow.then targets, cycles)
	if crossRefErrors := ValidateCrossTaskReferences(spec); len(crossRefErrors) > 0 {
		errors = append(errors, crossRefErrors...)
	}

	// Step 2b: Task config required field validation
	if configErrors := ValidateTaskConfigRequiredFields(spec); len(configErrors) > 0 {
		errors = append(errors, configErrors...)
	}

	// Step 2c: Model reference validation (harness-aware)
	if modelErrors := ValidateModelReferences(spec); len(modelErrors) > 0 {
		errors = append(errors, modelErrors...)
	}

	// Step 2d: Human-input timeout policy validation (fail closed on
	// policies the runtime cannot honor)
	if policyErrors := ValidateHumanInputTimeoutPolicies(spec); len(policyErrors) > 0 {
		errors = append(errors, policyErrors...)
	}

	// Step 3: Budget warnings
	if budgetWarnings := CheckBudgetWarnings(spec.Budget, spec.Tasks); len(budgetWarnings) > 0 {
		warnings = append(warnings, budgetWarnings...)
	}

	// Step 4: Expression warnings ($context.env.* → should be $env.*)
	if exprWarnings := CheckExpressionWarnings(spec); len(exprWarnings) > 0 {
		warnings = append(warnings, exprWarnings...)
	}

	if len(errors) > 0 {
		log.Warn().
			Int("errors", len(errors)).
			Int("warnings", len(warnings)).
			Msg("Validation failed (state: INVALID)")
		return &serverlessv1.ServerlessWorkflowValidation{
			State:       serverlessv1.ValidationState_INVALID,
			Yaml:        yaml,
			Errors:      errors,
			Warnings:    warnings,
			ValidatedAt: timestamppb.Now(),
		}, nil
	}

	log.Info().
		Int("warnings", len(warnings)).
		Msg("Validation passed (state: VALID)")

	return &serverlessv1.ServerlessWorkflowValidation{
		State:       serverlessv1.ValidationState_VALID,
		Yaml:        yaml,
		Errors:      errors,
		Warnings:    warnings,
		ValidatedAt: timestamppb.Now(),
	}, nil
}

File: backend/services/stigmer-server/src/domain/workflow/validation/validator.ts
/**
 * In-process workflow-spec validation — ports
 * pkg/domain/workflow/validation/validator.go. Validates a WorkflowSpec
 * entirely in-process: converts the spec to CNCF Serverless Workflow DSL
 * YAML, runs cross-reference validation, the declared task-config
 * constraints (#805), the surface rules, harness-aware model validation,
 * the human-input timeout-policy rule, and produces budget/expression
 * warnings.
 *
 * Verdict states:
 *   - VALID: conversion succeeded, no structural errors
 *   - INVALID: user error (bad structure, missing fields, bad references)
 *   - FAILED: system error (nil spec)
 * The yaml field always carries the generated CNCF YAML when conversion
 * succeeds (even for INVALID), as it aids debugging. A machinery fault
 * (the validation infrastructure itself failing) THROWS — callers surface
 * it as a gRPC internal error, mirroring Go's (nil, err) contract.
 */
import { create } from "@bufbuild/protobuf";
import { timestampNow } from "@bufbuild/protobuf/wkt";

import type { WorkflowSpec } from "@stigmer/protos/ai/stigmer/agentic/workflow/v1/spec_pb";
import {
  ServerlessWorkflowValidationSchema,
  ValidationState,
} from "@stigmer/protos/ai/stigmer/agentic/workflow/v1/serverless/validation_pb";
import type { ServerlessWorkflowValidation } from "@stigmer/protos/ai/stigmer/agentic/workflow/v1/serverless/validation_pb";

import type { Logger } from "../../../boot/logger.js";
import { protoToYaml } from "../converter/converter.js";
import type { ModelCatalogProvider } from "../registry/model-catalog-provider.js";
import { checkBudgetWarnings } from "./budget-warnings.js";
import {
  validateCrossTaskReferences,
  validateTaskConfigSurfaceRules,
  validateTaskKinds,
} from "./crossref.js";
import { checkExpressionWarnings } from "./expression-warnings.js";
import { validateHumanInputTimeoutPolicies } from "./human-input-validation.js";
import { validateModelReferences } from "./model-validation.js";
import { validateTaskConfigConstraints } from "./task-config-constraints.js";

export class InProcessValidator {
  constructor(
    private readonly modelRegistry: ModelCatalogProvider,
    private readonly logger: Logger,
  ) {}

  validate(spec: WorkflowSpec | undefined): ServerlessWorkflowValidation {
    if (spec === undefined) {
      return create(ServerlessWorkflowValidationSchema, {
        state: ValidationState.FAILED,
        errors: ["WorkflowSpec cannot be nil"],
        validatedAt: timestampNow(),
      });
    }

    this.logger.debug("starting in-process workflow validation");

    // Step 0: task kind validation (fail fast for unknown kinds).
    const kindErrors = validateTaskKinds(spec);
    if (kindErrors.length > 0) {
      return create(ServerlessWorkflowValidationSchema, {
        state: ValidationState.INVALID,
        errors: kindErrors,
        validatedAt: timestampNow(),
      });
    }

    // Step 1: convert proto to CNCF YAML.
    let yaml: string;
    try {
      yaml = protoToYaml(spec);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      this.logger.error("YAML generation failed", { error: message });
      return create(ServerlessWorkflowValidationSchema, {
        state: ValidationState.INVALID,
        errors: [`Failed to generate YAML: ${message}`],
        validatedAt: timestampNow(),
      });
    }

    this.logger.debug("YAML generation succeeded", { yamlLength: yaml.length });

    const errors: string[] = [];
    const warnings: string[] = [];

    // Step 2: cross-reference validation (unique names, flow.then targets,
    // cycles).
    errors.push(...validateCrossTaskReferences(spec));

    // Step 2b: typed task-config constraint validation — protovalidate over
    // each strict-unmarshaled config, nested tasks included (#805). A
    // throw here is a fault in the validation machinery, not a spec
    // problem, and propagates as a system error.
    errors.push(...validateTaskConfigConstraints(spec));

    // Step 2c: bespoke task-config surface rules (semantics beyond what
    // the config protos can declare).
    errors.push(...validateTaskConfigSurfaceRules(spec));

    // Step 2d: model reference validation (harness-aware).
    errors.push(...validateModelReferences(this.modelRegistry, spec));

    // Step 2e: human-input timeout policy validation (fail closed on
    // policies the runtime cannot honor).
    errors.push(...validateHumanInputTimeoutPolicies(spec));

    // Step 3: budget warnings.
    warnings.push(...checkBudgetWarnings(spec.budget, spec.tasks));

    // Step 4: expression warnings ($context.env.* → should be $env.*).
    warnings.push(...checkExpressionWarnings(spec));

    if (errors.length > 0) {
      this.logger.warn("validation failed (state: INVALID)", {
        errors: errors.length,
        warnings: warnings.length,
      });
      return create(ServerlessWorkflowValidationSchema, {
        state: ValidationState.INVALID,
        yaml,
        errors,
        warnings,
        validatedAt: timestampNow(),
      });
    }

    this.logger.info("validation passed (state: VALID)", {
      warnings: warnings.length,
    });

    return create(ServerlessWorkflowValidationSchema, {
      state: ValidationState.VALID,
      yaml,
      errors,
      warnings,
      validatedAt: timestampNow(),
    });
  }
}
